dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+   */
+  virtual bool ProcessOutputDesc(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<Backend> &ws, bool is_inferred) {
+    if (is_inferred && IsExpanding()) {
+      CoalesceOutputShapes(output_desc, ws);
+    }
+    return is_inferred;
+  }
+
+  virtual void ExpandArgument(const ArgumentWorkspace &ws, const std::string &arg_name,
+                              const TensorVector<CPUBackend> &arg_tensor) {
+    auto expand_idx = GetArgExpandDescIdx(arg_name);
+    auto expanded_arg = ExpandArgumentLike(arg_tensor, arg_name, expand_idx);
+    auto expanded_handle = std::make_shared<TensorVector<CPUBackend>>(std::move(expanded_arg));
+    AddArgumentInputHelper(arg_name, expanded_handle);
+  }
+
+  /**
+   * @brief Expands the tensor arguments from ``ws`` and adds them to expanded workspace.
+   * Arguments marked as supporting per-frame tensors and with leading `F` in tensor layout are
+   * unfolded (and additionally broadcasted if the channels in the operator's input are unfolded) to
+   * match the operator's input. If the argument is unfolded, each sample of the argument must have
+   * the number of frames equal to either: one (so it can be broadcasted) or to the number of frames
+   * in the corresponding sample of operator's expandable input.
+   *
+   * Tensor inputs not marked per-frame or with no leading frames in the layout are broadcasted to
+   * match the operator's expandable input.
+   *
+   * Assumes ``IsExpandble()`` is true, i.e. there is some expandable input to match the expanded
+   * arguments.
+   */
+  virtual void ExpandArguments(const ArgumentWorkspace &ws) {
+    for (const auto &arg_input : ws) {
+      auto &shared_tvec = arg_input.second.tvec;
+      assert(shared_tvec);
+      ExpandArgument(ws, arg_input.first, *shared_tvec);
+    }
+  }
+
+  /**
+   * @brief Gets the ExpandDesc instance that describes the expandable prefix of the layout and
+   * corresponding prefixes of the shapes in the batch. By default, the sequence shape and
+   * expandable layout are assumed to be the same for the output and the expandable input. If that's
+   * not the case, the method may be overriden to provide different ExpandDesc instance.
+   */
+  virtual const ExpandDesc &GetOutputExpandDesc(const workspace_t<Backend> &ws,
+                                                int output_idx) const {
+    (void)output_idx;
+    return GetInputExpandDesc(ExpandLikeIdx());
+  }
+
+  virtual int GetArgExpandDescIdx(const std::string &arg_name) const {
+    (void)arg_name;
+    return ExpandLikeIdx();
+  }
+
+  const ExpandDesc &GetInputExpandDesc(int input_idx) const {
+    DALI_ENFORCE(0 <= input_idx && static_cast<size_t>(input_idx) < input_expand_desc_.size());
+    return input_expand_desc_[input_idx];
+  }
+
+  int ExpandLikeIdx() const {
+    return expand_like_idx_;
+  }
+
+  void SetupSequenceOperator(const workspace_t<Backend> &ws) {
+    expand_like_idx_ = -1;
```

Comment:
I assumed they may change, as those out-of-the box fixed layouts and dims are not there yet. If we were to parse them once, we would have to, for now, check if they did not change anyway (not sure if that would make much of a difference performance-wise). ExpandDesc also process the shape, so if we were to somehow optimize the setup part, it would have to drop that part or be handled separately.